test/extensions/tracers/datadog/span_test.cc

```diff
@@ -0,0 +1,294 @@
+#include <vector>
+
+#include "absl/time/time.h"
+
+#include "source/common/tracing/null_span_impl.h"
+#include "source/extensions/tracers/datadog/span.h"
+#include "source/extensions/tracers/datadog/time_util.h"
+
+#include "test/mocks/tracing/mocks.h"
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+namespace datadog {
+namespace tracing {
+
+bool operator==(const TimePoint& left, const TimePoint& right) {
+  return left.wall == right.wall && left.tick == right.tick;
+}
+
+} // namespace tracing
+} // namespace datadog
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+namespace {
+
+class NullLogger : public datadog::tracing::Logger {
+public:
+  ~NullLogger() override = default;
+
+  void log_error(const LogFunc&) override {}
+  void log_startup(const LogFunc&) override {}
+
+  void log_error(const datadog::tracing::Error&) override {}
+  void log_error(datadog::tracing::StringView) override {}
+};
+
+struct MockCollector : public datadog::tracing::Collector {
+  datadog::tracing::Expected<void>
+  send(std::vector<std::unique_ptr<datadog::tracing::SpanData>>&& spans,
+       const std::shared_ptr<datadog::tracing::TraceSampler>&) override {
+    chunks.push_back(std::move(spans));
+    return {};
+  }
+
+  nlohmann::json config_json() const override {
+    return nlohmann::json::object({{"type", "Envoy::Extensions::Tracers::Datadog::MockCollector"}});
+  }
+
+  ~MockCollector() override = default;
+
+  std::vector<std::vector<std::unique_ptr<datadog::tracing::SpanData>>> chunks;
+};
+
+struct TestSetup {
+  const std::uint64_t id;
+  const std::shared_ptr<MockCollector> collector;
+  const datadog::tracing::TracerConfig config;
+  datadog::tracing::Tracer tracer;
+  datadog::tracing::Span span;
+
+  explicit TestSetup(std::uint64_t id = 0xcafebabe)
+      : id(id), collector(std::make_shared<MockCollector>()), config(makeConfig(collector)),
+        tracer(
+            // Override the tracer's ID generator so that all trace IDs and span
+            // IDs are `id`.
+            *datadog::tracing::finalize_config(config), [id]() { return id; },
+            datadog::tracing::default_clock),
+        span(tracer.create_span()) {}
+
+private:
+  static datadog::tracing::TracerConfig
+  makeConfig(const std::shared_ptr<datadog::tracing::Collector>& collector) {
+    datadog::tracing::TracerConfig config;
+    config.defaults.service = "testsvc";
+    config.collector = collector;
+    config.logger = std::make_shared<NullLogger>();
+    // Drop all spans. Equivalently, we could keep all spans.
+    datadog::tracing::TraceSamplerConfig::Rule rule;
+    rule.sample_rate = 0;
+    config.trace_sampler.rules.push_back(std::move(rule));
+    return config;
+  }
+};
+
+std::chrono::system_clock::time_point now() {
+  // 14:45 May 6th, 2010 in New York
+  absl::CivilSecond datetime{2010, 5, 3, 14, 45};
+  return absl::ToChronoTime(absl::FromCivil(datetime, absl::FixedTimeZone(-4 * 60 * 60)));
```

Comment:
hm.

The back-story here is we didn't used have all the absl time functions in Envoy at all, but we wanted to limit sleeps and other use of real time in tests, so we have a bunch of lint checks. The intent of the lint checks is to use timers and/or simulated time or mock time rather than sleeps and similar functions.

You avoid the lint checks with this, but I feel like this subverts the intent.

Why do we need all this? Can we use simulated time instead?

Reply:
`Span::spawnChild` and `Span::log` are part of the Envoy tracing interface that we must implement. Both have a parameter of type `SystemTime`. Any caller of one of those member functions must provide a `SystemTime` value.

For the purposes of these tests, the value of the `SystemTime` does not matter, so long as I know what it is. Initially I used the current time. Then I constructed a fixed time using `mktime`. Now I've constructed a fixed time using `absl::ToChronoTime`.

How could simulated time be used here?

Reply:
SimulatedTimeSystem has a method systemTime() which returns a SystemTime.

It gets initialized, by default, from the current real-time when the test is run, but it does not advance with real-time after that; it only advances due to methods called on the SimulatedTimeSystem to advance time.

So for your purposes you can just instantiate one and call systemTime() on it.

test/extensions/tracers/zipkin/zipkin_core_types_test.cc:118 is one usage that might be similar enough.

Reply:
Thanks, will do.